// cubelets/service/build.js
var util = require('util');
var events = require('events');
var request = require('request');
var url = require('url');
var fs = require('fs');
var temp = require('temp');
var config = require('../config.json');

var Build = function(buildID, program, cubelet) {
	this.id = buildID;
	this.program = program;
	this.cubelet = cubelet;
	this.hex = undefined;
};

var BuildService = function() {

	events.EventEmitter.call(this);

	var builds = {};
	var apiUrl = config['urls']['api'];
	var hexDataUrl = config['urls']['data_hex'];
	var service = this;

	var CompilerStatus = {
		COMPILING: 404,
		READY: 200
	};

	var requestCompilerStatus = function(buildID, callback) {
		request.get({
			url: apiUrl + '/compile-result/' + buildID + '/',
			json: true
		}, function(error, response, body) {
			if (error) {
				service.emit('error', error);
				return;
			}
			var status = body['status_code'];
			callback(status);
		});
	};

	var requestHexDownload = function(build) {
		request.get(hexDataUrl + build.id, function(error, response, body) {
			if (error) {
				console.error('Error downloading hex.');
				service.emit('error', error);
				return;
			}
			build.hex = body;
			service.emit('complete', build);
		});
	};

	var addBuild = function(buildID, program, cubelet) {
		return builds[buildID] = new Build(buildID, program, cubelet);
	};

	var removeBuild = function(buildID) {
		delete builds[buildID];
	};

	this.getBuilds = function() {
		return this.builds;
	};

	this.completeBuild = function(build) {
		removeBuild(build.id);
	};

	function emitBuildProgress(build, percent) {
		service.emit('progress', build, percent);
	};

	this.requestBuild = function(program, cubelet) {
		request.post({
			url: apiUrl + '/compile-request/',
			json: true,
			form: {
				'program_code': program.code,
				'program_title': program.name,
				'cube_id': cubelet.id
			}
		}, function(error, response, body) {
			if (error) {
				console.error('Request error.', error);
				service.emit('error', error);
				return;
			}
			if (response.statusCode != 200) {
				console.error('Bad response.', response);
				service.emit('error');
				return;
			}
			var statusCode = body['status_code'];
			if (statusCode != 200) {
				console.error('Bad response status.', body);
				service.emit('error');
				return;
			}
			var buildID = body['u'];
			var build = addBuild(buildID, program, cubelet);
			emitBuildProgress(build, 0);
			var intervalCount = 0;
			var intervalID = setInterval(function() {
				requestCompilerStatus(buildID, function(status) {
					if (++intervalCount > 100) {
						service.emit('error');
						clearInterval(intervalID);
						return;
					}
					switch (status) {
						case CompilerStatus.COMPILING:
							emitBuildProgress(build, 100.0 * intervalCount / 3); // This is a guess
							break;
						case CompilerStatus.READY:
							clearInterval(intervalID);
							requestHexDownload(build);
							break;
						default:
							clearInterval(intervalID);
							service.emit('error');
							break;
					}
				});
			}, 1 * 1000)
		});
	};

};

util.inherits(BuildService, events.EventEmitter);

module.exports = BuildService;
